src/snowflake/snowpark/dataframe.py

```diff
@@ -2237,19 +2244,32 @@ def unpivot(
             ---------------------------------------------
             <BLANKLINE>
         """
+        # AST.
```

Comment:
can you add `_emit_ast` as guard? We'll use that for rollout parameter.

Reply:
That approach will cost more than it saves now.

A feature flag:

- Starts out as "off" by default
- Early adopters can set it to "on", everyone else has it off
- At some point, the feature transitions to on by default
- Some users might hit blocking issues. They can set the feature flag to "off"
- At some point, all blocking issues are fixed for good.
- The feature flag disappears altogether, together with any fallback code paths that might have been present

`emit_ast` has semantics orthogonal to the rollout status of the feature. Using `emit_ast` for two purposes at once implies that:

- We introduce `emit_ast` in the signature of all public APIs, not just the ones called internally from other public APIs
- Whoever has to remove the "feature flag" uses of `emit_ast` has to differentiate those uses from the "silence internal uses of public APIs" uses of `emit_ast` in function signatures and in implementations. It is this last part that will be very expensive for whoever does it.

Reply:
I might have not been precise here: We'll add a decorator to guard the rollout of the IR feature for [SNOW-1557209](https://snowflakecomputing.atlassian.net/browse/SNOW-1557209), i.e. something like

```
@guard
def public_api(..., _emit_ast=True):
   ...
```
This guard will simply pass `_emit_ast=False` to the API. The behavior of the decorator `guard` will be controlled by a variable then (feature flag). For existing APIs, the sort-of established rule right now is that will have a parameter `_emit_ast:bool=True`.

However, I sense that you have some very good points here. Should we schedule a call with @sfc-gh-azwiegincew to align on the strategy with `_emit_ast`?

Reply:
The decorator is an implementation detail of the rollout. That's less interesting. The salient point is that we'd end up using `emit_ast` for two purposes, so (assuming we want to finish the rollout and retire the feature flag), untangling the two purposes at the end will result in non-trivial work, and in changes to the public API itself.

Reply:
Hm, how could we do better?

Reply:
One question I'd ask is "how do other Snowpark features get rolled out?" Is the decorator + parameter pattern common?

Another question would be - with or without a decorator, if `emit_ast` didn't exist, how would feature control work? I presume some piece of code would fetch configuration from one or more places (`.toml` file, environment variable, command-line flag, server response etc), make a policy decision and pass it around - by storing it in a global variable, or by returning it from a (memoized?) helper function etc.

In any case, I propose treating feature control completely separately from `emit_ast`.

For functions that don't have an `emit_ast` parameter, the logic would be:

```python
    if enable_snowpark_ir:
        ...
```

Functions that have an `emit_ast` parameter can do one of:

```python
    if enable_snowpark_ir:
        if _emit_ast:
            ...
```

This variant would be preferable if the two `if` statements result in distinct `else` clauses. If both `else` clauses do nothing, simpler variants will work.

```python
    if enable_snowpark_ir and _emit_ast:
        ...
```

```python
    _emit_ast = _emit_ast and enable_snowpark_ir
    if _emit_ast:
        ...
```

Reply:
Thanks for providing some more in-depth context here! I'd actually say it'd be good to discuss this in today's standup. For the scope of this PR, it sounds to me that we're aligned that adding a `_emit_ast` argument similar to the existing APIs would no matter which route we go for the feature flag be a good idea. I'd therefore suggest to for now simply use `_emit_ast` without a double-purpose in mind (i.e., simply to emit/not emit AST) and circle back on the feature flag in the standup. That would allow to land this PR, and not introduce other, new concepts with it. Wdyt? Is this a reasonable approach for this PR?

Reply:
Sure, I can do that.